### DISCUSSION TITLE
How can I query about code owners for a PR?

### DISCUSSION THREAD

**Discussion by chdsbd** [community/community#24702] [Apps, API and Webhooks]
<p>I’d like to fetch the code owners associated with a pull request or at least the repository.</p>
<p>Is this possible through existing APIs or does this feature not exist? My backup plan is to write code owner evaluation logic myself, but that seems like it shouldn’t be necessary.</p>

**Comment by wei** [comment]
<p>Through the <a href="https://developer.github.com/v3/pulls/#get-a-single-pull-request" rel="nofollow noopener">existing API</a>, you can fetch pull request author, repo owner of the source of the pull request, assignees, users that merges the pull request, and get commits with committers and authors. What are you looking for specifically?</p>

**Comment by chdsbd** [comment]
<p>I want to find the code owners requested for review on a specific PR. It doesn’t seem possible through the API.</p>

**Comment by wei** [comment]
<p><a href="https://help.github.com/en/articles/about-code-owners" rel="nofollow noopener">https://help.github.com/en/articles/about-code-owners</a></p>
<p>&gt; Code owners are automatically requested for review when someone opens a pull request that modifies code that they own.</p>
<p>So you could query requested reviewers.</p>

**Comment by chdsbd** [comment]
<p>I wanted to identify if a requested reviewer was requested as a code owner.</p>

**Comment by ThirdEyeRose** [comment]
<p>Hi <a class="mention" href="/chdsbd">@chdsbd</a>!</p>
<p>While it’s not possible to do this with a single API, you can create a script that allows you to crossreference the requested reviewers with the content of the CODEOWNERS file using the <a href="https://developer.github.com/v3/repos/contents/#get-contents" rel="nofollow noopener">Get Contents API endpoint</a>. Since the <a href="https://help.github.com/en/articles/about-code-owners#codeowners-file-location" rel="nofollow noopener">CODEOWNERS file has a limited number of locations that it can be located</a>, it should be easy to find with a couple of calls.</p>
<p>Hope that helps!</p>

**Comment by chdsbd** [comment]
<p>Thanks. That’s what I’ve settled on. The challenge I see is that a user could be in the code owners file but not necessarily be requested because of their ownership. Another user could request their review for code they don’t own.</p>

**Comment by ThirdEyeRose** [comment]
<p><a class="mention" href="/chdsbd">@chdsbd</a>, that’s very true. You’d probably have to do some advanced checking of the files in the PR and how they compare to the CODEOWNERS file as well to be sure. </p>

**Comment by payneChen** [comment]
<p><a class="mention" href="/that-pat">@that-pat</a>, we have similiar request to check whether the approval is coming from code owner or not, does GitHub have any plan to expose one single API for code owner check?</p>

**Comment by nethgato** [comment, marked as answer]
<p>Hi <a class="mention" href="/paynechen">@payneChen</a> o/</p>
<p>Since this OP, there have been some internal goings-on, with relation to CODEOWNERS data being available from our APIs.</p>
<p>Within the <code>PullRequest</code> object for GraphQL, we have the <code>reviewDecision</code> field, which you can see, here:</p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/free-pro-team@latest/graphql/reference/objects#pullrequest" target="_blank" rel="noopener">docs.github.com</a>
  </header>
  <article class="onebox-body">
    <img src="" class="thumbnail" width="16" height="16">

<h3><a href="https://docs.github.com/en/free-pro-team@latest/graphql/reference/objects#pullrequest" target="_blank" rel="noopener">Objects - GitHub Docs</a></h3>



  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Continuing with GraphQL, we also have the <code>PullRequestReviewDecision</code> enum, which includes the status values for review requests:</p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/free-pro-team@latest/graphql/reference/enums#pullrequestreviewdecision" target="_blank" rel="noopener">docs.github.com</a>
  </header>
  <article class="onebox-body">
    <img src="" class="thumbnail" width="16" height="16">

<h3><a href="https://docs.github.com/en/free-pro-team@latest/graphql/reference/enums#pullrequestreviewdecision" target="_blank" rel="noopener">Enums - GitHub Docs</a></h3>



  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>As far as I’m aware (currently) we have no REST functionality, or a specific CODEOWNER endpoint as a single source for this information.</p>
<p>Very curious to know if the above solves for your use case. If it doesn’t, I would highly recommend submitting your feedback, through this form:</p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <a href="https://support.github.com/contact/feedback" target="_blank" rel="noopener">support.github.com</a>
  </header>
  <article class="onebox-body">
    <img src="" class="thumbnail" width="16" height="16">

<h3><a href="https://support.github.com/contact/feedback" target="_blank" rel="noopener">Share feedback - GitHub Support</a></h3>



  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>…which gets reviewed directly by our PMs.</p>
<p>Sincerely hope this helps! :bow:</p>

**Comment by chdsbd** [comment]
<p>It seems like <code>reviewDecision</code> should work for my use case, but there’s a bug with the field that’s preventing me from using it.</p>
<p>The <code>reviewDecision</code> field causes the GitHub GraphQL API to error when “Require review from Code Owners” is enabled  and the pull request has a file matching CODEOWNERS. I have a ticket open with GitHub Support about this.</p>

**Comment by nethgato** [comment]
<p>Hey <a class="mention" href="/chdsbd">@chdsbd</a> thanks so much for voicing your experience. Very curious to know how your exchange goes with that internal support ticket.</p>
<p>If anything seems relevant for the Community, would you mind posting back?</p>
<p>Anyone else reading is also empowered to @ mention me and I can see if there was traction that might be suited for a broader audience, than the 1:1 ticket mentioned above.</p>
<p>:bow:</p>

**Comment by jd** [comment]
<p><a class="mention" href="/nethgato">@nethgato</a> I’m pinging you because we’d really need to have code owner reviews exposed for the REST API too. Is there anything we can help with?</p>

**Comment by nethgato** [comment]
<p>Hi <a class="mention" href="/jd">@jd</a> o/</p>
<p>As-is, the best thing you could do, is to add your voice to those who would like to see similar data in REST, as we have in GraphQL. The most helpful path forward, would be to submit via our feedback form, here:</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://support-assets.github.com/assets/favicon-3dcd2c977a91b2e3f396b300c5ff0d572295ca632ddb7ba1adcacc1ea68dd5c0.ico" class="site-icon" width="16" height="16">
      <a href="https://support.github.com/contact/feedback" target="_blank" rel="noopener">support.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://support.github.com/contact/feedback" target="_blank" rel="noopener">Share feedback - GitHub Support</a></h3>

<p><a href="https://support.github.com/contact/feedback" target="_blank" rel="noopener">//support.github.com/contact/feedback</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by KaranErry** [comment]
<p>This is the best way I’ve found to do this:</p>
<ol>
<li>List requested reviewers for a pull request using the following API: <a href="https://docs.github.com/en/github-ae@latest/rest/reference/pulls#list-requested-reviewers-for-a-pull-request" rel="noopener nofollow ugc">https://docs.github.com/en/github-ae@latest/rest/reference/pulls#list-requested-reviewers-for-a-pull-request</a>. This will give you a list of usernames and a list of teams.</li>
<li>Codeowners are essentially just teams or individuals assigned to watch over certain parts of the codebase. In my case, I wanted a list of usernames who are part of each codeowning team. For this I used the Teams API: <a href="https://docs.github.com/en/rest/reference/teams#list-team-members" rel="noopener nofollow ugc">https://docs.github.com/en/rest/reference/teams#list-team-members</a>.</li>
</ol>
<p>Caveats of this approach:</p>
<ul>
<li>Step 1 of this approach lists out all of those whose reviews were requested on the PR. While Github automatically requests reviews from codeowning individuals and teams, one can manually request reviews as well. This step does not distinguish between automatic review requests (which suggest codeowner status) and manual review requests.</li>
</ul>

**Comment by rewanthtammana** [comment]
@KaranErry There are a few problems with the solution you proposed.

1. The reviewers endpoint - it will ONLY list the names of users you manually request someone a review. CODEOWNERS isn't in this list.
2. Listing members of CODEOWNERS will work only in few cases. It will not work/help in many.

For example, see the below snippet of CODEOWNERS file

  ```
      // .github/CODEOWNERS
      * @random/team1 @random/team2
      docs/ @random/team2 @random/team3
   ```
     
If the PR is raised for some file inside `docs/`, then it allows approvals only `team2` & `team3` approval. If the PR contains changes elsewhere, then it requests review from `team1` & `team2` only. The CODEOWNERS endpoint you provided will not be able to help us in identifying the specific team information for every PR.

The GraphQL solution shared by @nethgato (https://github.com/orgs/community/discussions/24702#discussioncomment-3245124) might help but having a REST API is definitely solves the problem. Not sure why Github isn't exposing this endpoint despite many requests.

Some special features related to CODEOWNERS are provided by Github to the cloud version but they aren't existing in on-prem/free version of github.

**Reply by rewanthtammana** [reply]
Update: Just tried the solution provided by @nethgato, it isn't helpful or enables us to print the list of reviewers for a specific PR.

**Comment by Greesb** [comment]
Any news on this ? It would be really helpful to be able to know the list of codeowners required to approve a specific PR. I believe it can just be as simple as a user having a special team in the list of reviewers of a PR.

**Reply by kryocentric** [reply]
A hack I've been using -

1. get a list of code reviewers
2. remove code reviewers one by one
3. wait 10 seconds for github's side to sync
4. get a list of code reviewers again. code owners can't be removed, so people who survived the purge are code owners.
5. add people back if you have to

down side - ugly history of removing/adding reviewers

**Comment by jeremycarroll** [comment]
I see the information _is_ available, but not in the API .... e.g.
```
curl https://github.com/dotnet/samples/deferred-metadata/main/machine-learning/models/sentimentanalysis/sentiment_model.zip
```
gives
```
{
  "showCodeNavSurvey": false,
  "showLicenseMeta": false,
  "license": null,
  "newIssuePath": "/dotnet/samples/issues/new",
  "newDiscussionPath": null,
  "codeownerInfo": {
    "codeownerPath": "/dotnet/samples/blob/main/.github/CODEOWNERS#L21",
    "ownedByCurrentUser": false,
    "ownersForFile": "@luisquintanilla",
    "ruleForPathLine": 21
  }
}
```
When used on a private repo authentication has to be done using a `user_session` cookie, an API token does not work.
Simply allowing auth using an API token here would be a step forward, although I can understand that a more principled addition to the API would be better.

**Comment by joh-klein** [comment]
Because of this discussion, I wrote a GitHub Action that outputs the affected code owners of a pull request:

```
name: 'Check Affected Code Owners'

on:
  pull_request:
    types: [opened, synchronize]

jobs:
  affected-codeowners:
    runs-on: ubuntu-latest
    permissions:
      contents: read
      pull-requests: read

    steps:
      - name: Check out repository
        uses: actions/checkout@v4

      - name: List affected code owners
        uses: dytab/affected-codeowners@v1
        id: affected-codeowners

      - name: Output affected code owners
        run: |
          echo "Individual Owners: ${{ steps.affected-codeowners.outputs.individual-owners }}"
          echo "Grouped Owners: ${{ steps.affected-codeowners.outputs.grouped-owners }}"
```

https://github.com/marketplace/actions/affected-code-owners
